File: SetupWizardMacApp/Sources/SetupWizard/Services/SetupStateManager.swift
import Foundation

/// Manages persistent setup state across wizard sessions
actor SetupStateManager {
    private var state: SetupStateFile
    private let stateFilePath: String
    
    init() {
        let home = FileManager.default.homeDirectoryForCurrentUser
        let wineDir = home.appendingPathComponent(".wine")
        let stateFilePath = wineDir.appendingPathComponent("setup-state.json").path
        
        // Create wine directory if needed
        try? FileManager.default.createDirectory(at: wineDir, withIntermediateDirectories: true)
        
        self.stateFilePath = stateFilePath
        
        // Load existing state or create new
        if let loaded = SetupStateFile.load(from: stateFilePath) {
            self.state = loaded
            Logger.shared.debug("Loaded existing setup state")
        } else {
            self.state = SetupStateFile()
            Logger.shared.debug("Created new setup state")
        }
    }
    
    func getCurrentStep() -> SetupStateFile.Step {
        return state.currentStep
    }
    
    func setCurrentStep(_ step: SetupStateFile.Step) async {
        state.currentStep = step
        try? state.save(to: stateFilePath)
    }
    
    func setSelectedFolder(_ folder: String) async {
        state.selectedFolder = folder
        try? state.save(to: stateFilePath)
    }
    
    func getSelectedFolder() -> String? {
        return state.selectedFolder
    }
    
    func setPatchedExePath(_ path: String) async {
        state.patchedExePath = path
        try? state.save(to: stateFilePath)
    }
    
    func getPatchedExePath() -> String? {
        return state.patchedExePath
    }
    
    func setSelectedWineType(_ type: String) async {
        state.selectedWineType = type
        try? state.save(to: stateFilePath)
    }
    
    func getSelectedWineType() -> String? {
        return state.selectedWineType
    }
    
    func setDotNetStatus(_ status: SetupStateFile.OperationStatus, error: String? = nil) async {
        state.dotNetStatus = status
        state.dotNetError = error
        try? state.save(to: stateFilePath)
    }
    
    func getDotNetStatus() -> (status: SetupStateFile.OperationStatus, error: String?) {
        return (state.dotNetStatus, state.dotNetError)
    }
    
    func setCurrentOperationPID(_ pid: Int32) async {
        state.currentOperationPID = pid
        state.operationStartTime = Date()
        try? state.save(to: stateFilePath)
    }
    
    func clearCurrentOperationPID() async {
        state.currentOperationPID = nil
        state.operationStartTime = nil
        try? state.save(to: stateFilePath)
    }
    
    func recordCancellation() async {
        state.lastCancellationTime = Date()
        try? state.save(to: stateFilePath)
    }
    
    func getLastCancellationTime() -> Date? {
        return state.lastCancellationTime
    }
    
    func markComplete() async {
        state.wizardStatus = .completed
        state.completionTime = Date()
        try? state.save(to: stateFilePath)
    }
    
    func markFailed(error: String) async {
        state.wizardStatus = .failed
        state.lastError = error
        try? state.save(to: stateFilePath)
    }
    
    func getWizardStatus() -> (status: SetupStateFile.OperationStatus, error: String?) {
        return (state.wizardStatus, state.lastError)
    }
    
    func reset() async {
        state = SetupStateFile()
        try? FileManager.default.removeItem(atPath: stateFilePath)
        Logger.shared.debug("Reset setup state")
    }
    
    func getFullState() -> SetupStateFile {
        return state
    }
    
    func cleanupStaleProcesses() async {
        if let pid = state.currentOperationPID, state.dotNetStatus == .inProgress {
            // Check if process still exists
            let killResult = kill(pid, 0) // Signal 0 just checks if process exists
            if killResult == -1 {
                Logger.shared.log("Process \(pid) no longer exists, cleaning up")
                state.currentOperationPID = nil
                state.dotNetStatus = .failed
                state.dotNetError = "Process was terminated unexpectedly"
                try? state.save(to: stateFilePath)
            }
        }
    }
}

File: SetupWizardCore/Services/SetupStateManager.cs
using SetupWizardCore.Models;

namespace SetupWizardCore.Services;

/// <summary>
/// Manages persistent setup state across wizard sessions.
/// Port of Swift SetupStateManager actor.
/// </summary>
public class SetupStateManager
{
    private SetupState _state;
    private readonly string _stateFilePath;
    private readonly object _lock = new();

    public SetupStateManager()
    {
        _stateFilePath = SetupState.GetStateFilePath();

        // Load existing state or create new
        var loaded = SetupState.Load(_stateFilePath);
        if (loaded != null)
        {
            _state = loaded;
            SetupWizardLogger.Instance.Debug("Loaded existing setup state");
        }
        else
        {
            _state = new SetupState();
            SetupWizardLogger.Instance.Debug("Created new setup state");
        }
    }

    public SetupState.Step GetCurrentStep() => _state.CurrentStep;

    public void SetCurrentStep(SetupState.Step step)
    {
        lock (_lock)
        {
            _state.CurrentStep = step;
            Save();
        }
    }

    public void SetSelectedFolder(string folder)
    {
        lock (_lock)
        {
            _state.SelectedFolder = folder;
            Save();
        }
    }

    public string? GetSelectedFolder() => _state.SelectedFolder;

    public void SetPatchedExePath(string path)
    {
        lock (_lock)
        {
            _state.PatchedExePath = path;
            Save();
        }
    }

    public string? GetPatchedExePath() => _state.PatchedExePath;

    public void SetSelectedWineType(string type)
    {
        lock (_lock)
        {
            _state.SelectedWineType = type;
            Save();
        }
    }

    public string? GetSelectedWineType() => _state.SelectedWineType;

    public void SetDotNetStatus(SetupState.OperationStatus status, string? error = null)
    {
        lock (_lock)
        {
            _state.DotNetStatus = status;
            _state.DotNetError = error;
            Save();
        }
    }

    public (SetupState.OperationStatus status, string? error) GetDotNetStatus()
    {
        lock (_lock)
        {
            return (_state.DotNetStatus, _state.DotNetError);
        }
    }

    public void SetCurrentOperationPID(int pid)
    {
        lock (_lock)
        {
            _state.CurrentOperationPID = pid;
            _state.OperationStartTime = DateTime.UtcNow;
            Save();
        }
    }

    public void ClearCurrentOperationPID()
    {
        lock (_lock)
        {
            _state.CurrentOperationPID = null;
            _state.OperationStartTime = null;
            Save();
        }
    }

    public void RecordCancellation()
    {
        lock (_lock)
        {
            _state.LastCancellationTime = DateTime.UtcNow;
            Save();
        }
    }

    public DateTime? GetLastCancellationTime()
    {
        lock (_lock) return _state.LastCancellationTime;
    }

    public void MarkComplete()
    {
        lock (_lock)
        {
            _state.WizardStatus = SetupState.OperationStatus.Completed;
            _state.CompletionTime = DateTime.UtcNow;
            Save();
        }
    }

    public void MarkFailed(string error)
    {
        lock (_lock)
        {
            _state.WizardStatus = SetupState.OperationStatus.Failed;
            _state.LastError = error;
            Save();
        }
    }

    public (SetupState.OperationStatus status, string? error) GetWizardStatus()
    {
        lock (_lock) return (_state.WizardStatus, _state.LastError);
    }

    public void Reset()
    {
        lock (_lock)
        {
            _state = new SetupState();
            try { File.Delete(_stateFilePath); }
            catch { /* ignore */ }
            SetupWizardLogger.Instance.Debug("Reset setup state");
        }
    }

    public SetupState GetFullState()
    {
        lock (_lock) return _state;
    }

    /// <summary>
    /// Check if a previously recorded operation PID is still running; mark failed if stale.
    /// Port of Swift cleanupStaleProcesses().
    /// </summary>
    public void CleanupStaleProcesses()
    {
        lock (_lock)
        {
            if (_state.CurrentOperationPID.HasValue && _state.DotNetStatus == SetupState.OperationStatus.InProgress)
            {
                try
                {
                    var process = System.Diagnostics.Process.GetProcessById(_state.CurrentOperationPID!.Value);
                    if (process.HasExited)
                    {
                        SetupWizardLogger.Instance.Log("Process " + _state.CurrentOperationPID.Value + " no longer exists, cleaning up");
                        _state.CurrentOperationPID = null;
                        _state.DotNetStatus = SetupState.OperationStatus.Failed;
                        _state.DotNetError = "Process was terminated unexpectedly";
                        Save();
                    }
                }
                catch (ArgumentException)
                {
                    // Process doesn't exist
                    SetupWizardLogger.Instance.Log("Process " + _state.CurrentOperationPID!.Value + " no longer exists, cleaning up");
                    _state.CurrentOperationPID = null;
                    _state.DotNetStatus = SetupState.OperationStatus.Failed;
                    _state.DotNetError = "Process was terminated unexpectedly";
                    Save();
                }
            }
        }
    }

    private void Save()
    {
        try
        {
            _state.Save(_stateFilePath);
        }
        catch (Exception ex)
        {
            SetupWizardLogger.Instance.Error("Failed to save state: " + ex.Message);
        }
    }
}
